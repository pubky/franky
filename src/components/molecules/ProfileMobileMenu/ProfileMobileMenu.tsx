'use client';

import Link from 'next/link';
import { usePathname } from 'next/navigation';
import * as Libs from '@/libs';
import { PROFILE_MENU_ITEMS, getProfileCount, ProfileCounts } from '@/molecules/ProfileMenu/profileMenu.shared';
import { PROFILE_ROUTES } from '@/app';

export interface ProfileMobileMenuProps {
  className?: string;
  counts?: ProfileCounts;
}

export function ProfileMobileMenu({ className, counts }: ProfileMobileMenuProps) {
  const pathname = usePathname();

  const getCount = (label: string) => getProfileCount(label, counts);

  const isSelected = (path: string) => pathname === path;

  // Reorder menu items for mobile: Tagged first, then others
  const taggedItem = PROFILE_MENU_ITEMS.find((item) => item.path === PROFILE_ROUTES.TAGGED);
  const otherItems = PROFILE_MENU_ITEMS.filter((item) => item.path !== PROFILE_ROUTES.TAGGED);
  const mobileMenuItems = taggedItem ? [taggedItem, ...otherItems] : PROFILE_MENU_ITEMS;

  return (
    <div
      className={Libs.cn('fixed top-18 left-0 right-0 z-20 bg-background shadow-xl border-b border-border', className)}
    >
      <div className="flex w-full items-center">
        {/* Menu items - Tagged is first on mobile with CircleUserRound icon */}
        {mobileMenuItems.map((item) => {
          // Use CircleUserRound icon for Tagged on mobile, otherwise use the item's icon
          const Icon = item.path === PROFILE_ROUTES.TAGGED ? Libs.CircleUserRound : item.icon;
          const selected = isSelected(item.path);
          const count = getCount(item.label);

          return (
            <Link
              key={item.path}
              href={item.path}
              className={Libs.cn(
                'flex flex-1 justify-center items-center gap-2 px-0 py-1.5 transition-all relative border-b',
                selected
                  ? 'text-foreground border-b border-white'
                  : 'text-muted-foreground border-b border-transparent',
              )}
              aria-label={item.label}
              aria-current={selected ? 'page' : undefined}
            >
              <div className="flex items-center justify-center gap-2 px-2.5 py-2 rounded-md">
                <Icon size={20} className="shrink-0" />
                {count !== undefined && <span className="text-sm font-medium leading-normal">{count}</span>}
              </div>
            </Link>
          );
        })}
      </div>
    </div>
  );
}
