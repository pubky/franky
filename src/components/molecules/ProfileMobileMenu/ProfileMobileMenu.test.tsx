import { describe, it, expect, vi, beforeEach } from 'vitest';
import { render, screen } from '@testing-library/react';
import { usePathname } from 'next/navigation';
import { ProfileMobileMenu } from './ProfileMobileMenu';

// Mock next/navigation
vi.mock('next/navigation', () => ({
  usePathname: vi.fn(() => '/profile/notifications'),
}));

describe('ProfileMobileMenu', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('renders all mobile menu items', () => {
    render(<ProfileMobileMenu />);

    // Should render 7 links for 7 menu items (Tagged is first)
    const links = screen.getAllByRole('link');
    expect(links).toHaveLength(7);

    // Check that all menu items are present by their aria-labels
    expect(screen.getByLabelText('Tagged')).toBeInTheDocument();
    expect(screen.getByLabelText('Notifications')).toBeInTheDocument();
    expect(screen.getByLabelText('Posts')).toBeInTheDocument();
    expect(screen.getByLabelText('Replies')).toBeInTheDocument();
    expect(screen.getByLabelText('Followers')).toBeInTheDocument();
    expect(screen.getByLabelText('Following')).toBeInTheDocument();
    expect(screen.getByLabelText('Friends')).toBeInTheDocument();
  });

  it('displays counts when provided', () => {
    const counts = {
      notifications: 3,
      posts: 10,
      replies: 5,
      followers: 100,
      following: 50,
      friends: 20,
      tagged: 3,
    };
    render(<ProfileMobileMenu counts={counts} />);

    // Check for the count numbers in span elements
    expect(screen.getAllByText('3')).toHaveLength(2); // notifications and tagged both have count 3
    expect(screen.getByText('10')).toBeInTheDocument();
    expect(screen.getByText('5')).toBeInTheDocument();
    expect(screen.getByText('100')).toBeInTheDocument();
    expect(screen.getByText('50')).toBeInTheDocument();
    expect(screen.getByText('20')).toBeInTheDocument();
  });

  it('handles navigation when menu items are clicked', () => {
    render(<ProfileMobileMenu />);

    const postsLink = screen.getByLabelText('Posts');
    expect(postsLink).toHaveAttribute('href', '/profile/posts');
  });

  it('highlights active menu item based on current pathname', () => {
    // Test with posts route
    vi.mocked(usePathname).mockReturnValue('/profile/posts');
    render(<ProfileMobileMenu />);

    // Posts should be active (though we need to check implementation details)
    expect(screen.getByLabelText('Posts')).toBeInTheDocument();
  });

  it('renders without counts when no counts provided', () => {
    render(<ProfileMobileMenu />);

    // Should still render all menu items but without counts
    expect(screen.getByLabelText('Posts')).toBeInTheDocument();
    expect(screen.getByLabelText('Replies')).toBeInTheDocument();
    expect(screen.getByLabelText('Tagged')).toBeInTheDocument();
    expect(screen.getByLabelText('Followers')).toBeInTheDocument();
    expect(screen.getByLabelText('Following')).toBeInTheDocument();
    expect(screen.getByLabelText('Friends')).toBeInTheDocument();

    // Should not have count text
    expect(screen.queryByText(/\(.*\)/)).not.toBeInTheDocument();
  });

  it('handles all navigation routes correctly', () => {
    render(<ProfileMobileMenu />);

    // Verify all links have correct href attributes
    expect(screen.getByLabelText('Tagged')).toHaveAttribute('href', '/profile/tagged');
    expect(screen.getByLabelText('Notifications')).toHaveAttribute('href', '/profile/notifications');
    expect(screen.getByLabelText('Replies')).toHaveAttribute('href', '/profile/replies');
    expect(screen.getByLabelText('Followers')).toHaveAttribute('href', '/profile/followers');
    expect(screen.getByLabelText('Following')).toHaveAttribute('href', '/profile/following');
    expect(screen.getByLabelText('Friends')).toHaveAttribute('href', '/profile/friends');
    expect(screen.getByLabelText('Posts')).toHaveAttribute('href', '/profile/posts');
  });

  it('matches snapshot', () => {
    const { container } = render(<ProfileMobileMenu />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot with counts', () => {
    const counts = {
      posts: 10,
      replies: 5,
      tagged: 3,
      followers: 100,
      following: 50,
      friends: 20,
    };
    const { container } = render(<ProfileMobileMenu counts={counts} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
