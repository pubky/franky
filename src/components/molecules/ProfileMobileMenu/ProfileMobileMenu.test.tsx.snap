// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProfileMobileMenu > matches snapshot 1`] = `
<div
  class="fixed top-18 left-0 right-0 z-20 bg-background shadow-xl border-b border-border"
>
  <div
    class="flex w-full items-center"
  >
    <a
      aria-label="Tagged"
      class="flex flex-1 justify-center items-center gap-2 px-0 py-1.5 transition-all relative text-muted-foreground border-b border-transparent"
      href="/profile/tagged"
    >
      <div
        class="flex items-center justify-center gap-2 px-2.5 py-2 rounded-md"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-circle-user-round shrink-0"
          fill="none"
          height="20"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M18 20a6 6 0 0 0-12 0"
          />
          <circle
            cx="12"
            cy="10"
            r="4"
          />
          <circle
            cx="12"
            cy="12"
            r="10"
          />
        </svg>
      </div>
    </a>
    <a
      aria-label="Notifications"
      class="flex flex-1 justify-center items-center gap-2 px-0 py-1.5 transition-all relative text-muted-foreground border-b border-transparent"
      href="/profile/notifications"
    >
      <div
        class="flex items-center justify-center gap-2 px-2.5 py-2 rounded-md"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-bell shrink-0"
          fill="none"
          height="20"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M10.268 21a2 2 0 0 0 3.464 0"
          />
          <path
            d="M3.262 15.326A1 1 0 0 0 4 17h16a1 1 0 0 0 .74-1.673C19.41 13.956 18 12.499 18 8A6 6 0 0 0 6 8c0 4.499-1.411 5.956-2.738 7.326"
          />
        </svg>
      </div>
    </a>
    <a
      aria-current="page"
      aria-label="Posts"
      class="flex flex-1 justify-center items-center gap-2 px-0 py-1.5 transition-all relative text-foreground border-b border-white"
      href="/profile/posts"
    >
      <div
        class="flex items-center justify-center gap-2 px-2.5 py-2 rounded-md"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-sticky-note shrink-0"
          fill="none"
          height="20"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M16 3H5a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2V8Z"
          />
          <path
            d="M15 3v4a2 2 0 0 0 2 2h4"
          />
        </svg>
      </div>
    </a>
    <a
      aria-label="Replies"
      class="flex flex-1 justify-center items-center gap-2 px-0 py-1.5 transition-all relative text-muted-foreground border-b border-transparent"
      href="/profile/replies"
    >
      <div
        class="flex items-center justify-center gap-2 px-2.5 py-2 rounded-md"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-message-circle shrink-0"
          fill="none"
          height="20"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M7.9 20A9 9 0 1 0 4 16.1L2 22Z"
          />
        </svg>
      </div>
    </a>
    <a
      aria-label="Followers"
      class="flex flex-1 justify-center items-center gap-2 px-0 py-1.5 transition-all relative text-muted-foreground border-b border-transparent"
      href="/profile/followers"
    >
      <div
        class="flex items-center justify-center gap-2 px-2.5 py-2 rounded-md"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-users-round shrink-0"
          fill="none"
          height="20"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M18 21a8 8 0 0 0-16 0"
          />
          <circle
            cx="10"
            cy="8"
            r="5"
          />
          <path
            d="M22 20c0-3.37-2-6.5-4-8a5 5 0 0 0-.45-8.3"
          />
        </svg>
      </div>
    </a>
    <a
      aria-label="Following"
      class="flex flex-1 justify-center items-center gap-2 px-0 py-1.5 transition-all relative text-muted-foreground border-b border-transparent"
      href="/profile/following"
    >
      <div
        class="flex items-center justify-center gap-2 px-2.5 py-2 rounded-md"
      >
        <svg
          class="shrink-0"
          fill="none"
          height="20"
          viewBox="0 0 20 20"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5.00016 17.5C5.00016 15.7319 5.70254 14.0362 6.95278 12.786C8.20303 11.5357 9.89872 10.8333 11.6668 10.8333M11.6668 10.8333C13.4349 10.8333 15.1306 11.5357 16.3809 12.786C17.6311 14.0362 18.3335 15.7319 18.3335 17.5M11.6668 10.8333C9.36564 10.8333 7.50016 8.96785 7.50016 6.66667C7.50016 4.36548 9.36564 2.5 11.6668 2.5C13.968 2.5 15.8335 4.36548 15.8335 6.66667C15.8335 8.96785 13.968 10.8333 11.6668 10.8333ZM1.66679 16.6666C1.66679 13.8583 3.33346 11.25 5.00012 9.99996C4.45227 9.58893 4.01419 9.0492 3.72465 8.42852C3.43511 7.80784 3.30303 7.12535 3.34011 6.44146C3.37719 5.75757 3.58227 5.09336 3.93723 4.50763C4.29218 3.92189 4.78605 3.43268 5.37512 3.08329"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
        </svg>
      </div>
    </a>
    <a
      aria-label="Friends"
      class="flex flex-1 justify-center items-center gap-2 px-0 py-1.5 transition-all relative text-muted-foreground border-b border-transparent"
      href="/profile/friends"
    >
      <div
        class="flex items-center justify-center gap-2 px-2.5 py-2 rounded-md"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-heart-handshake shrink-0"
          fill="none"
          height="20"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 14c1.49-1.46 3-3.21 3-5.5A5.5 5.5 0 0 0 16.5 3c-1.76 0-3 .5-4.5 2-1.5-1.5-2.74-2-4.5-2A5.5 5.5 0 0 0 2 8.5c0 2.3 1.5 4.05 3 5.5l7 7Z"
          />
          <path
            d="M12 5 9.04 7.96a2.17 2.17 0 0 0 0 3.08c.82.82 2.13.85 3 .07l2.07-1.9a2.82 2.82 0 0 1 3.79 0l2.96 2.66"
          />
          <path
            d="m18 15-2-2"
          />
          <path
            d="m15 18-2-2"
          />
        </svg>
      </div>
    </a>
  </div>
</div>
`;

exports[`ProfileMobileMenu > matches snapshot with counts 1`] = `
<div
  class="fixed top-18 left-0 right-0 z-20 bg-background shadow-xl border-b border-border"
>
  <div
    class="flex w-full items-center"
  >
    <a
      aria-label="Tagged"
      class="flex flex-1 justify-center items-center gap-2 px-0 py-1.5 transition-all relative text-muted-foreground border-b border-transparent"
      href="/profile/tagged"
    >
      <div
        class="flex items-center justify-center gap-2 px-2.5 py-2 rounded-md"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-circle-user-round shrink-0"
          fill="none"
          height="20"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M18 20a6 6 0 0 0-12 0"
          />
          <circle
            cx="12"
            cy="10"
            r="4"
          />
          <circle
            cx="12"
            cy="12"
            r="10"
          />
        </svg>
        <span
          class="text-sm font-medium leading-normal"
        >
          3
        </span>
      </div>
    </a>
    <a
      aria-label="Notifications"
      class="flex flex-1 justify-center items-center gap-2 px-0 py-1.5 transition-all relative text-muted-foreground border-b border-transparent"
      href="/profile/notifications"
    >
      <div
        class="flex items-center justify-center gap-2 px-2.5 py-2 rounded-md"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-bell shrink-0"
          fill="none"
          height="20"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M10.268 21a2 2 0 0 0 3.464 0"
          />
          <path
            d="M3.262 15.326A1 1 0 0 0 4 17h16a1 1 0 0 0 .74-1.673C19.41 13.956 18 12.499 18 8A6 6 0 0 0 6 8c0 4.499-1.411 5.956-2.738 7.326"
          />
        </svg>
      </div>
    </a>
    <a
      aria-current="page"
      aria-label="Posts"
      class="flex flex-1 justify-center items-center gap-2 px-0 py-1.5 transition-all relative text-foreground border-b border-white"
      href="/profile/posts"
    >
      <div
        class="flex items-center justify-center gap-2 px-2.5 py-2 rounded-md"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-sticky-note shrink-0"
          fill="none"
          height="20"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M16 3H5a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2V8Z"
          />
          <path
            d="M15 3v4a2 2 0 0 0 2 2h4"
          />
        </svg>
        <span
          class="text-sm font-medium leading-normal"
        >
          10
        </span>
      </div>
    </a>
    <a
      aria-label="Replies"
      class="flex flex-1 justify-center items-center gap-2 px-0 py-1.5 transition-all relative text-muted-foreground border-b border-transparent"
      href="/profile/replies"
    >
      <div
        class="flex items-center justify-center gap-2 px-2.5 py-2 rounded-md"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-message-circle shrink-0"
          fill="none"
          height="20"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M7.9 20A9 9 0 1 0 4 16.1L2 22Z"
          />
        </svg>
        <span
          class="text-sm font-medium leading-normal"
        >
          5
        </span>
      </div>
    </a>
    <a
      aria-label="Followers"
      class="flex flex-1 justify-center items-center gap-2 px-0 py-1.5 transition-all relative text-muted-foreground border-b border-transparent"
      href="/profile/followers"
    >
      <div
        class="flex items-center justify-center gap-2 px-2.5 py-2 rounded-md"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-users-round shrink-0"
          fill="none"
          height="20"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M18 21a8 8 0 0 0-16 0"
          />
          <circle
            cx="10"
            cy="8"
            r="5"
          />
          <path
            d="M22 20c0-3.37-2-6.5-4-8a5 5 0 0 0-.45-8.3"
          />
        </svg>
        <span
          class="text-sm font-medium leading-normal"
        >
          100
        </span>
      </div>
    </a>
    <a
      aria-label="Following"
      class="flex flex-1 justify-center items-center gap-2 px-0 py-1.5 transition-all relative text-muted-foreground border-b border-transparent"
      href="/profile/following"
    >
      <div
        class="flex items-center justify-center gap-2 px-2.5 py-2 rounded-md"
      >
        <svg
          class="shrink-0"
          fill="none"
          height="20"
          viewBox="0 0 20 20"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5.00016 17.5C5.00016 15.7319 5.70254 14.0362 6.95278 12.786C8.20303 11.5357 9.89872 10.8333 11.6668 10.8333M11.6668 10.8333C13.4349 10.8333 15.1306 11.5357 16.3809 12.786C17.6311 14.0362 18.3335 15.7319 18.3335 17.5M11.6668 10.8333C9.36564 10.8333 7.50016 8.96785 7.50016 6.66667C7.50016 4.36548 9.36564 2.5 11.6668 2.5C13.968 2.5 15.8335 4.36548 15.8335 6.66667C15.8335 8.96785 13.968 10.8333 11.6668 10.8333ZM1.66679 16.6666C1.66679 13.8583 3.33346 11.25 5.00012 9.99996C4.45227 9.58893 4.01419 9.0492 3.72465 8.42852C3.43511 7.80784 3.30303 7.12535 3.34011 6.44146C3.37719 5.75757 3.58227 5.09336 3.93723 4.50763C4.29218 3.92189 4.78605 3.43268 5.37512 3.08329"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
          />
        </svg>
        <span
          class="text-sm font-medium leading-normal"
        >
          50
        </span>
      </div>
    </a>
    <a
      aria-label="Friends"
      class="flex flex-1 justify-center items-center gap-2 px-0 py-1.5 transition-all relative text-muted-foreground border-b border-transparent"
      href="/profile/friends"
    >
      <div
        class="flex items-center justify-center gap-2 px-2.5 py-2 rounded-md"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-heart-handshake shrink-0"
          fill="none"
          height="20"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 14c1.49-1.46 3-3.21 3-5.5A5.5 5.5 0 0 0 16.5 3c-1.76 0-3 .5-4.5 2-1.5-1.5-2.74-2-4.5-2A5.5 5.5 0 0 0 2 8.5c0 2.3 1.5 4.05 3 5.5l7 7Z"
          />
          <path
            d="M12 5 9.04 7.96a2.17 2.17 0 0 0 0 3.08c.82.82 2.13.85 3 .07l2.07-1.9a2.82 2.82 0 0 1 3.79 0l2.96 2.66"
          />
          <path
            d="m18 15-2-2"
          />
          <path
            d="m15 18-2-2"
          />
        </svg>
        <span
          class="text-sm font-medium leading-normal"
        >
          20
        </span>
      </div>
    </a>
  </div>
</div>
`;
